File: backend/api-date/java/src/DateService.java
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class DateService {
    private static final Map<String, String[]> LOCATIONS = new LinkedHashMap<>();

    static {
        put("us", "United States", "America/New_York");
        put("ca", "Canada", "America/Toronto");
        put("br", "Brazil", "America/Sao_Paulo");
        put("ar", "Argentina", "America/Argentina/Buenos_Aires");
        put("cl", "Chile", "America/Santiago");
        put("co", "Colombia", "America/Bogota");
        put("pe", "Peru", "America/Lima");
        put("gb", "United Kingdom", "Europe/London");
        put("ie", "Ireland", "Europe/Dublin");
        put("fr", "France", "Europe/Paris");
        put("de", "Germany", "Europe/Berlin");
        put("es", "Spain", "Europe/Madrid");
        put("it", "Italy", "Europe/Rome");
        put("nl", "Netherlands", "Europe/Amsterdam");
        put("be", "Belgium", "Europe/Brussels");
        put("ch", "Switzerland", "Europe/Zurich");
        put("pt", "Portugal", "Europe/Lisbon");
        put("se", "Sweden", "Europe/Stockholm");
        put("no", "Norway", "Europe/Oslo");
        put("fi", "Finland", "Europe/Helsinki");
        put("pl", "Poland", "Europe/Warsaw");
        put("ua", "Ukraine", "Europe/Kyiv");
        put("tr", "Turkey", "Europe/Istanbul");
        put("ru", "Russia", "Europe/Moscow");
        put("sa", "Saudi Arabia", "Asia/Riyadh");
        put("ae", "United Arab Emirates", "Asia/Dubai");
        put("eg", "Egypt", "Africa/Cairo");
        put("za", "South Africa", "Africa/Johannesburg");
        put("ng", "Nigeria", "Africa/Lagos");
        put("ke", "Kenya", "Africa/Nairobi");
        put("in", "India", "Asia/Kolkata");
        put("pk", "Pakistan", "Asia/Karachi");
        put("bd", "Bangladesh", "Asia/Dhaka");
        put("cn", "China", "Asia/Shanghai");
        put("jp", "Japan", "Asia/Tokyo");
        put("kr", "South Korea", "Asia/Seoul");
        put("sg", "Singapore", "Asia/Singapore");
        put("id", "Indonesia", "Asia/Jakarta");
        put("th", "Thailand", "Asia/Bangkok");
        put("vn", "Vietnam", "Asia/Ho_Chi_Minh");
        put("ph", "Philippines", "Asia/Manila");
        put("my", "Malaysia", "Asia/Kuala_Lumpur");
        put("au", "Australia", "Australia/Sydney");
        put("nz", "New Zealand", "Pacific/Auckland");
        put("il", "Israel", "Asia/Jerusalem");
        put("ir", "Iran", "Asia/Tehran");
        put("ma", "Morocco", "Africa/Casablanca");
        put("dz", "Algeria", "Africa/Algiers");
        put("et", "Ethiopia", "Africa/Addis_Ababa");
        put("gh", "Ghana", "Africa/Accra");
        put("mx-central", "Mexico Central", "America/Mexico_City");
        put("mx-northwest", "Mexico Northwest", "America/Tijuana");
        put("mx-southeast", "Mexico Southeast", "America/Cancun");
    }

    private DateService() {
    }

    public static Map<String, Object> buildLocationsPayload() {
        List<Map<String, Object>> locations = new ArrayList<>();
        for (Map.Entry<String, String[]> entry : LOCATIONS.entrySet()) {
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("location", entry.getKey());
            item.put("label", entry.getValue()[0]);
            item.put("timezone", entry.getValue()[1]);
            locations.add(item);
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("location_count", locations.size());
        payload.put("locations", locations);
        return payload;
    }

    public static Map<String, Object> buildTimePayload(String location, String ip) {
        String normalized = normalizeLocation(location);
        String[] metadata = LOCATIONS.get(normalized);
        ZonedDateTime localNow = ZonedDateTime.now(ZoneId.of(metadata[1]));
        ZonedDateTime utcNow = ZonedDateTime.now(ZoneId.of("UTC"));

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("location", normalized);
        payload.put("label", metadata[0]);
        payload.put("timezone", metadata[1]);
        payload.put("local_time", localNow.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));
        payload.put("utc_offset", localNow.format(DateTimeFormatter.ofPattern("XX")));
        payload.put("utc_time", utcNow.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss 'UTC+00:00'")));
        payload.put("ip", ip);
        return payload;
    }

    public static Map<String, Object> buildResourcesPayload() {
        List<Map<String, Object>> resources = new ArrayList<>();
        resources.add(resource("date://auth-reference", "auth-reference", "Resumen de autenticación requerida por el servicio date."));
        resources.add(resource("date://location-reference", "location-reference", "Listado de ubicaciones soportadas y sus zonas horarias."));
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("resources", resources);
        return payload;
    }

    public static Map<String, Object> buildResourceContentsPayload(String resourceUri) {
        String text;
        if ("date://auth-reference".equals(resourceUri)) {
            text = String.join("\n",
                "Date API authentication reference",
                "- Header Authorization: Bearer <token>",
                "- Header X-Date-Client: <client-id>",
                "- Endpoints require authentication for GET requests"
            );
        } else if ("date://location-reference".equals(resourceUri)) {
            StringBuilder builder = new StringBuilder();
            boolean first = true;
            for (Map.Entry<String, String[]> entry : LOCATIONS.entrySet()) {
                if (!first) {
                    builder.append('\n');
                }
                first = false;
                builder.append("- ")
                    .append(entry.getKey())
                    .append(": ")
                    .append(entry.getValue()[0])
                    .append(" (")
                    .append(entry.getValue()[1])
                    .append(")");
            }
            text = builder.toString();
        } else {
            throw new IllegalArgumentException(resourceUri);
        }

        Map<String, Object> content = new LinkedHashMap<>();
        content.put("uri", resourceUri);
        content.put("mimeType", "text/plain");
        content.put("text", text);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("contents", List.of(content));
        return payload;
    }

    public static Map<String, Object> buildPromptsPayload() {
        List<Map<String, Object>> prompts = new ArrayList<>();
        prompts.add(prompt(
            "single-location-time",
            "Pide la hora actual de una ubicación concreta usando get_current_time.",
            List.of(argument("location", "Código de ubicación soportado.", true))
        ));
        prompts.add(prompt(
            "compare-locations",
            "Pide comparar la hora actual entre dos ubicaciones.",
            List.of(
                argument("from_location", "Primera ubicación.", true),
                argument("to_location", "Segunda ubicación.", true)
            )
        ));
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("prompts", prompts);
        return payload;
    }

    public static Map<String, Object> buildPromptDetailsPayload(String promptName, String fromLocation, String toLocation) {
        if ("single-location-time".equals(promptName)) {
            String location = normalizeLocation(fromLocation == null ? "mx-central" : fromLocation);
            return promptDetails(
                "Prompt para obtener la hora de una sola ubicación.",
                List.of(
                    message("system", "Usa la herramienta get_current_time y responde con la hora obtenida."),
                    message("user", "Dime la hora actual para " + location + " usando get_current_time.")
                )
            );
        }
        if ("compare-locations".equals(promptName)) {
            String first = normalizeLocation(fromLocation == null ? "mx-central" : fromLocation);
            String second = normalizeLocation(toLocation == null ? "us" : toLocation);
            return promptDetails(
                "Prompt para comparar dos ubicaciones usando get_current_time.",
                List.of(
                    message("system", "Usa get_current_time para ambas ubicaciones y compara los resultados."),
                    message("user", "Compara la hora actual entre " + first + " y " + second + " usando get_current_time para cada ubicación.")
                )
            );
        }
        throw new IllegalArgumentException(promptName);
    }

    public static String normalizeLocation(String location) {
        if (location == null || location.isBlank()) {
            return "mx-central";
        }
        String normalized = location.trim().toLowerCase();
        if (!LOCATIONS.containsKey(normalized)) {
            throw new IllegalArgumentException(location);
        }
        return normalized;
    }

    private static void put(String key, String label, String zoneId) {
        LOCATIONS.put(key, new String[] {label, zoneId});
    }

    private static Map<String, Object> resource(String uri, String name, String description) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("uri", uri);
        payload.put("name", name);
        payload.put("description", description);
        payload.put("mimeType", "text/plain");
        return payload;
    }

    private static Map<String, Object> prompt(String name, String description, List<Map<String, Object>> arguments) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("name", name);
        payload.put("description", description);
        payload.put("arguments", arguments);
        return payload;
    }

    private static Map<String, Object> argument(String name, String description, boolean required) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("name", name);
        payload.put("description", description);
        payload.put("required", required);
        return payload;
    }

    private static Map<String, Object> promptDetails(String description, List<Map<String, Object>> messages) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("description", description);
        payload.put("messages", messages);
        return payload;
    }

    private static Map<String, Object> message(String role, String text) {
        Map<String, Object> content = new LinkedHashMap<>();
        content.put("type", "text");
        content.put("text", text);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("role", role);
        payload.put("content", content);
        return payload;
    }
}

File: backend/api-date/python/date_service.py
from __future__ import annotations

from datetime import datetime, timezone
from zoneinfo import ZoneInfo


DATE_LOCATIONS = {
    "us": ("United States", "America/New_York"),
    "ca": ("Canada", "America/Toronto"),
    "br": ("Brazil", "America/Sao_Paulo"),
    "ar": ("Argentina", "America/Argentina/Buenos_Aires"),
    "cl": ("Chile", "America/Santiago"),
    "co": ("Colombia", "America/Bogota"),
    "pe": ("Peru", "America/Lima"),
    "gb": ("United Kingdom", "Europe/London"),
    "ie": ("Ireland", "Europe/Dublin"),
    "fr": ("France", "Europe/Paris"),
    "de": ("Germany", "Europe/Berlin"),
    "es": ("Spain", "Europe/Madrid"),
    "it": ("Italy", "Europe/Rome"),
    "nl": ("Netherlands", "Europe/Amsterdam"),
    "be": ("Belgium", "Europe/Brussels"),
    "ch": ("Switzerland", "Europe/Zurich"),
    "pt": ("Portugal", "Europe/Lisbon"),
    "se": ("Sweden", "Europe/Stockholm"),
    "no": ("Norway", "Europe/Oslo"),
    "fi": ("Finland", "Europe/Helsinki"),
    "pl": ("Poland", "Europe/Warsaw"),
    "ua": ("Ukraine", "Europe/Kyiv"),
    "tr": ("Turkey", "Europe/Istanbul"),
    "ru": ("Russia", "Europe/Moscow"),
    "sa": ("Saudi Arabia", "Asia/Riyadh"),
    "ae": ("United Arab Emirates", "Asia/Dubai"),
    "eg": ("Egypt", "Africa/Cairo"),
    "za": ("South Africa", "Africa/Johannesburg"),
    "ng": ("Nigeria", "Africa/Lagos"),
    "ke": ("Kenya", "Africa/Nairobi"),
    "in": ("India", "Asia/Kolkata"),
    "pk": ("Pakistan", "Asia/Karachi"),
    "bd": ("Bangladesh", "Asia/Dhaka"),
    "cn": ("China", "Asia/Shanghai"),
    "jp": ("Japan", "Asia/Tokyo"),
    "kr": ("South Korea", "Asia/Seoul"),
    "sg": ("Singapore", "Asia/Singapore"),
    "id": ("Indonesia", "Asia/Jakarta"),
    "th": ("Thailand", "Asia/Bangkok"),
    "vn": ("Vietnam", "Asia/Ho_Chi_Minh"),
    "ph": ("Philippines", "Asia/Manila"),
    "my": ("Malaysia", "Asia/Kuala_Lumpur"),
    "au": ("Australia", "Australia/Sydney"),
    "nz": ("New Zealand", "Pacific/Auckland"),
    "il": ("Israel", "Asia/Jerusalem"),
    "ir": ("Iran", "Asia/Tehran"),
    "ma": ("Morocco", "Africa/Casablanca"),
    "dz": ("Algeria", "Africa/Algiers"),
    "et": ("Ethiopia", "Africa/Addis_Ababa"),
    "gh": ("Ghana", "Africa/Accra"),
    "mx-central": ("Mexico Central", "America/Mexico_City"),
    "mx-northwest": ("Mexico Northwest", "America/Tijuana"),
    "mx-southeast": ("Mexico Southeast", "America/Cancun"),
}


def build_locations_payload() -> dict[str, object]:
    locations = [
        {
            "location": code,
            "label": label,
            "timezone": timezone_name,
        }
        for code, (label, timezone_name) in sorted(DATE_LOCATIONS.items())
    ]
    return {
        "location_count": len(locations),
        "locations": locations,
    }


def build_time_payload(location: str, client_ip: str) -> dict[str, object]:
    normalized = normalize_location(location)
    label, timezone_name = DATE_LOCATIONS[normalized]
    local_now = datetime.now(ZoneInfo(timezone_name))
    utc_now = datetime.now(timezone.utc)
    return {
        "location": normalized,
        "label": label,
        "timezone": timezone_name,
        "local_time": local_now.strftime("%Y-%m-%d %H:%M:%S"),
        "utc_offset": local_now.strftime("%z"),
        "utc_time": utc_now.strftime("%Y-%m-%d %H:%M:%S UTC+00:00"),
        "ip": client_ip,
    }


def build_resources_payload() -> dict[str, object]:
    return {
        "resources": [
            {
                "uri": "date://auth-reference",
                "name": "auth-reference",
                "description": "Resumen de autenticación requerida por el servicio date.",
                "mimeType": "text/plain",
            },
            {
                "uri": "date://location-reference",
                "name": "location-reference",
                "description": "Listado de ubicaciones soportadas y sus zonas horarias.",
                "mimeType": "text/plain",
            },
        ]
    }


def build_resource_contents_payload(resource_uri: str) -> dict[str, object]:
    if resource_uri == "date://auth-reference":
        text = "\n".join(
            [
                "Date API authentication reference",
                "- Header Authorization: Bearer <token>",
                "- Header X-Date-Client: <client-id>",
                "- Endpoints require authentication for GET requests",
            ]
        )
    elif resource_uri == "date://location-reference":
        text = "\n".join(
            f"- {code}: {label} ({timezone_name})"
            for code, (label, timezone_name) in sorted(DATE_LOCATIONS.items())
        )
    else:
        raise KeyError(resource_uri)

    return {
        "contents": [
            {
                "uri": resource_uri,
                "mimeType": "text/plain",
                "text": text,
            }
        ]
    }


def build_prompts_payload() -> dict[str, object]:
    return {
        "prompts": [
            {
                "name": "single-location-time",
                "description": "Pide la hora actual de una ubicación concreta usando get_current_time.",
                "arguments": [
                    {"name": "location", "description": "Código de ubicación soportado.", "required": True},
                ],
            },
            {
                "name": "compare-locations",
                "description": "Pide comparar la hora actual entre dos ubicaciones.",
                "arguments": [
                    {"name": "from_location", "description": "Primera ubicación.", "required": True},
                    {"name": "to_location", "description": "Segunda ubicación.", "required": True},
                ],
            },
        ]
    }


def build_prompt_details_payload(prompt_name: str, from_location: str | None = None, to_location: str | None = None) -> dict[str, object]:
    if prompt_name == "single-location-time":
        resolved_location = normalize_location(from_location or "mx-central")
        return {
            "description": "Prompt para obtener la hora de una sola ubicación.",
            "messages": [
                {
                    "role": "system",
                    "content": {
                        "type": "text",
                        "text": "Usa la herramienta get_current_time y responde con la hora obtenida.",
                    },
                },
                {
                    "role": "user",
                    "content": {
                        "type": "text",
                        "text": f"Dime la hora actual para {resolved_location} usando get_current_time.",
                    },
                },
            ],
        }

    if prompt_name == "compare-locations":
        resolved_from = normalize_location(from_location or "mx-central")
        resolved_to = normalize_location(to_location or "us")
        return {
            "description": "Prompt para comparar dos ubicaciones usando get_current_time.",
            "messages": [
                {
                    "role": "system",
                    "content": {
                        "type": "text",
                        "text": "Usa get_current_time para ambas ubicaciones y compara los resultados.",
                    },
                },
                {
                    "role": "user",
                    "content": {
                        "type": "text",
                        "text": (
                            f"Compara la hora actual entre {resolved_from} y {resolved_to} "
                            "usando get_current_time para cada ubicación."
                        ),
                    },
                },
            ],
        }

    raise KeyError(prompt_name)


def normalize_location(location: str | None) -> str:
    if not location:
        return "mx-central"
    normalized = location.strip().lower()
    if normalized in DATE_LOCATIONS:
        return normalized
    raise KeyError(location)
